ddl/table_lock.go

```diff
@@ -0,0 +1,226 @@
+			return i
+		}
+	}
+	return -1
+}
+
+// checkAndLockTable uses to check table locked and acquire the table lock for the request session.
+func checkAndLockTable(tbInfo *model.TableInfo, idx int, arg *lockTablesArg) error {
+	if tbInfo.Lock == nil || len(tbInfo.Lock.Sessions) == 0 {
+		tbInfo.Lock = &model.TableLockInfo{
+			Tp: arg.LockTables[idx].Tp,
+		}
+		tbInfo.Lock.Sessions = append(tbInfo.Lock.Sessions, arg.SessionInfo)
+		return nil
+	}
+	if tbInfo.Lock.State == model.TableLockStatePreLock {
+		return nil
+	}
+	if tbInfo.Lock.Tp == model.TableLockRead && arg.LockTables[idx].Tp == model.TableLockRead {
+		sessionIndex := indexOfLockHolder(tbInfo.Lock.Sessions, arg.SessionInfo)
+		// repeat lock.
+		if sessionIndex >= 0 {
+			return nil
+		}
+		tbInfo.Lock.Sessions = append(tbInfo.Lock.Sessions, arg.SessionInfo)
+		return nil
+	}
+	// Unlock tables should execute before lock tables, so arg.SessionInfo in tbInfo.Lock.Sessions here is impossible.
+
+	return infoschema.ErrTableLocked.GenWithStackByArgs(tbInfo.Name.L, tbInfo.Lock.Tp, tbInfo.Lock.Sessions[0])
+}
+
+// checkTableLocked uses to check whether table was locked.
+func checkTableLocked(tbInfo *model.TableInfo, lockTp model.TableLockType, sessionInfo model.SessionInfo) error {
+	if tbInfo.Lock == nil || len(tbInfo.Lock.Sessions) == 0 {
+		return nil
+	}
+	if tbInfo.Lock.State == model.TableLockStatePreLock {
+		return nil
+	}
+	if tbInfo.Lock.Tp == model.TableLockRead && lockTp == model.TableLockRead {
+		return nil
+	}
+	sessionIndex := indexOfLockHolder(tbInfo.Lock.Sessions, sessionInfo)
+	// repeat lock.
+	if sessionIndex >= 0 {
+		if tbInfo.Lock.Tp == lockTp {
+			return nil
+		}
+		if len(tbInfo.Lock.Sessions) == 1 {
+			return nil
+		}
+	}
+	return infoschema.ErrTableLocked.GenWithStackByArgs(tbInfo.Name.L, tbInfo.Lock.Tp, tbInfo.Lock.Sessions[0])
+}
+
+// unlockTables uses unlock a batch of table lock one by one.
+func unlockTables(t *meta.Meta, job *model.Job, arg *lockTablesArg) (ver int64, err error) {
+	if arg.IndexOfUnlock >= len(arg.UnlockTables) {
+		return ver, nil
+	}
+	job.SchemaID = arg.UnlockTables[arg.IndexOfUnlock].SchemaID
+	job.TableID = arg.UnlockTables[arg.IndexOfUnlock].TableID
+	tbInfo, err := getTableInfo(t, job.TableID, job.SchemaID)
+	if err != nil {
+		if infoschema.ErrDatabaseNotExists.Equal(err) || infoschema.ErrTableNotExists.Equal(err) {
+			// The table maybe has been dropped. just ignore this err and go on.
+			arg.IndexOfUnlock++
+			job.Args = []interface{}{arg}
+			return ver, nil
+		} else {
+			job.State = model.JobStateCancelled
+			return ver, err
+		}
+	}
+
+	err = unlockTable(tbInfo, arg)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	ver, err = updateVersionAndTableInfo(t, job, tbInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	arg.IndexOfUnlock++
+	job.Args = []interface{}{arg}
+	job.SchemaState = model.StateDeleteOnly
```

Comment:
a better name?

Reply:
Any advice?

Reply:
How about add a new state `model.StateLocking` in parser?

Reply:
I think that would fine

Reply:
em... After reconsideration, I think It is redundant here, I will remove this.